lib/std/tar.zig

```diff
@@ -1,172 +1,1086 @@
+    // also, maybe this logic should be moved elsewhere. maybe 'fn header()'
+    const atime = if (header.atime == -1) time.nanoTimestamp() else header.atime;
+    const mtime = if (header.mtime == -1) time.nanoTimestamp() else header.mtime;
+    if (is_windows)
+        // workaround for 'panic: integer cast truncated bits' from
+        // file.updateTimes()
+        try file.updateTimes(truncateTime(atime), truncateTime(mtime))
+    else
+        try file.updateTimes(atime, mtime);
+
+    if (options.mode_mode == .executable_bit_only) {
+        if (std.fs.has_executable_bit) {
+            var modebits = std.StaticBitSet(32){
+                // .mask = @intCast(u32, header.mode),
+                // TODO - not sure using file.mode() is correct but it seems to
+                //        match gnu tar behavior on linux while using
+                //        header.mode does not
+                .mask = @intCast(u32, try file.mode()),
+            };
+            // copy the user exe bit to the group and other exe bits
+            // these bit indices count from the right:
+            //   u   g   o
+            //   rwx rwx rwx
+            //   876_543_210
+            // 0b000_000_000
+            const has_owner_exe_bit = modebits.isSet(6);
+            modebits.setValue(3, has_owner_exe_bit);
+            modebits.setValue(0, has_owner_exe_bit);
+            std.log.debug("mode old={o} new={o}", .{ header.mode, modebits.mask });
+            try file.chmod(modebits.mask);
+        }
+    }
+}
+
+fn setDirProperties(dir: fs.Dir, header: Header, options: Options) !void {
+    return setFileProperties(fs.File{ .handle = dir.fd }, header, options);
+}
+
+fn makePath(dir: fs.Dir, dir_name: []const u8, header: Header, options: Options) !void {
+    // openIterableDir() avoids .BADF => unreachable, // always a race condition
+    var subdir = try dir.makeOpenPathIterable(dir_name, .{});
+    defer subdir.close();
+    try setDirProperties(subdir.dir, header, options);
+}
+
+pub const Options = struct {
+    /// Number of directory levels to skip when extracting files.
+    strip_components: u32 = 0,
+    /// How to handle the "mode" property of files from within the tar file.
+    mode_mode: ModeMode = .executable_bit_only,
+
+    const ModeMode = enum {
+        /// The mode from the tar file is completely ignored. Files are created
+        /// with the default mode when creating files.
+        ignore,
+        /// The mode from the tar file is inspected for the owner executable bit
+        /// only. This bit is copied to the group and other executable bits.
+        /// Other bits of the mode are left as the default when creating files.
+        executable_bit_only,
+    };
+};
+
+pub fn pipeToFileSystem(dir: fs.Dir, reader: anytype, options: Options) !void {
+    var format = FormatSet.initMany(&.{ .ustar, .pax, .gnu });
+    var buf: [block_len * 2]u8 = undefined;
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
```

Comment:
This is certainly not a good idea. Allocators exist so that dynamic allocations like this aren't hidden.

Reply:
how does adding a `allocator: ?mem.Allocator` field to `Options` sound? maybe it shouldn't be optional?

Reply:
Oh, just realized that you did the same thing with pax headers, they may be more than 512 bytes and will need to be allocated. Better off making it required.

Reply:
ended up just adding an allocator param to pipeToFileSystem()